accepted/future-releases/extension-types/feature-specification.md

```diff
@@ -826,9 +829,18 @@ _is the extension type_
 <code>V\<T<sub>1</sub>, .. T<sub>s</sub>&gt;</code>,
 and that its static type _is an extension type_.
 
+We say that an extension type is _unrelated to `Future`_ in the case where
+it does not implement `Future<U>` for any `U` *(this means that it has no
+direct or indirect superinterface of the form `Future<...>`)*.
+
 It is a compile-time error if `await e` occurs, and the static type of
-`e` is an extension type which is not a subtype of `Future<T>` for any
-`T`.
+`e` satisfiers at least one of the following criteria:
+
+- an extension type which is unrelated to `Future`.
+- a type of the form `T?` or `FutureOr<T>` where `T` is a type that matches
+  one of the items in this list.
```

Comment:
If we give the property of being in the list a name, then we can refer to it recursively here.

Say:
> ```md
> We say that a type `S` is *incompatible with `await`* if:
> * `S` is an extension type which is unrelated to `Future`,
> * `S` is `T?` and `T` is incompatible with `await`, or
> * `S` is `T` bounded, where `T` is not `S`, and `T` is incompatible with `await`.
>
> An expression `await e` is a compile-time error if the static type of `e` is incompatible with `await`.
> ```

(IIRC, a type is trivially bounded by itself, which I guess is why you use the "one of the earlier items" phrase to avoid infinite regress. I guess calling it a coinductive definition would allow the infinite recursion.)